ocaml/xenopsd/xc/device.ml

```diff
@@ -1264,13 +1264,80 @@ module Varstored = SystemdDaemonMgmt (struct
   let pid_path domid = sprintf "/local/domain/%d/varstored-pid" domid
 end)
 
-module Swtpm = SystemdDaemonMgmt (struct
-  let name = "swtpm-wrapper"
+(* TODO: struct and include and uri to uri mapper, etc.
+   also xapi needs default backend set
+*)
+module Swtpm = struct
+  module D = SystemdDaemonMgmt (struct
+    let name = "swtpm-wrapper"
 
-  let use_pidfile = false
+    let use_pidfile = false
 
-  let pid_path domid = sprintf "/local/domain/%d/varstored-pid" domid
-end)
+    let pid_path domid = sprintf "/local/domain/%d/varstored-pid" domid
+  end)
+
+  let xs_path ~domid = Device_common.get_private_path domid ^ "/vtpm"
+
+  let vtpms_of_domid ~xs ~domid =
+    try
+      let vtpm_path = xs_path domid in
+      xs.Xs.directory vtpm_path
+      |> List.map @@ fun index ->
+         xs.Xs.read @@ Filename.concat vtpm_path index
+         |> Uuidm.of_string
+         |> Option.get
+    with _ -> []
+
+  let state_path =
+    (* for easier compat with dir:// mode, but can be anything.
+       If we implement VDI state storage this could be a block device
+    *)
+    Xenops_sandbox.Chroot.Path.of_string ~relative:"tpm2-00.permall"
+
+  let restore ~xs ~domid ~vm_uuid state =
+    if String.length state > 0 then begin
+      let path = Xenops_sandbox.Swtpm_guard.create ~domid ~vm_uuid state_path in
+      debug "Restored vTPM for domid %d: %d bytes, digest %s" domid (String.length state) (state |> Digest.string |> Digest.to_hex);
+      Unixext.write_string_to_file path state
+    end else
+      debug "vTPM state for domid %d is empty: not restoring" domid
+
+  let start_daemon dbg ~xs ~path ~args ~domid ~vm_uuid ~vtpm_uuid ~index () =
+    let state =
+      Varstore_privileged_client.Client.vtpm_get_contents dbg vtpm_uuid
+      |> Base64.decode_exn
+    in
+    let chroot = Xenops_sandbox.Swtpm_guard.chroot ~domid ~vm_uuid in
```

Comment:
The root of the sandbox depends on the domid and vm_uuid. Expecting the sandbox to exist to recover state seems suspect to me instead of simply deleting the sandbox when we need to.

Reply:
Yes, we probably want a way to check whether a file exists in the sandbox without necessarily creating the sandbox (which would create the file if it doesn't exist). I'll see whether I can plumb the Chroot.t parameter through.